From 3ecfceb9c2b4783b5e5af21ed30e04f0a2c80d14 Mon Sep 17 00:00:00 2001
From: Balamurugan Selvarajan <quic_bselvara@quicinc.com>
Date: Wed, 14 Sep 2022 14:44:13 +0530
Subject: [PATCH] ath12k: Enable PPE in fast Rx path

Submit the SKB to PPE module when PPE is enabled. This eventually improves
CPU idle percentage by 7 - 10%.

Signed-off-by: Balamurugan Selvarajan <quic_bselvara@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/dp_rx.c | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/drivers/net/wireless/ath/ath12k/dp_rx.c b/drivers/net/wireless/ath/ath12k/dp_rx.c
index 9782330..70e1466 100644
--- a/drivers/net/wireless/ath/ath12k/dp_rx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_rx.c
@@ -20,6 +20,11 @@
 #include "peer.h"
 #include "dp_mon.h"
 #include "hif.h"
+#ifdef CPTCFG_MAC80211_PPE_SUPPORT
+#include <ppe_vp_public.h>
+#include <ppe_vp_tx.h>
+#endif
+
 
 #define ATH12K_DP_RX_FRAGMENT_TIMEOUT_MS (2 * HZ)
 
@@ -2415,6 +2420,9 @@ static void ath12k_dp_rx_h_mpdu(struct ath12k *ar,
 	struct ieee80211_hdr *hdr;
 	struct ath12k_peer *peer;
 	u32 err_bitmap;
+#ifdef CPTCFG_MAC80211_PPE_SUPPORT
+	u32 vp;
+#endif
 	struct wireless_dev *wdev = NULL;
 	struct ath12k_sta *arsta = NULL;
 
@@ -2437,9 +2445,18 @@ static void ath12k_dp_rx_h_mpdu(struct ath12k *ar,
 		    ath12k_dp_rx_check_fast_rx(ar, msdu, rx_desc, peer)) {
 		        wdev = ieee80211_vif_to_wdev(peer->vif);
 		        if (wdev) {
+#ifdef CPTCFG_MAC80211_PPE_SUPPORT
+				vp = peer->vif->ppe_vp_num;
+#endif
 				spin_unlock_bh(&ar->ab->base_lock);
 				ath12k_dp_rx_h_csum_offload(ar, msdu);
 				msdu->dev = wdev->netdev;
+#ifdef CPTCFG_MAC80211_PPE_SUPPORT
+				if (vp) {
+					if (likely(ppe_vp_tx_to_ppe(vp, msdu)))
+						return;
+				}
+#endif
 				msdu->protocol = eth_type_trans(msdu, msdu->dev);
 				napi_gro_receive(rxcb->napi, msdu);
 				if (peer->sta)
-- 
2.17.1

